perception/traffic_light_estimator/src/node.cpp

```diff
@@ -0,0 +1,333 @@
+
+TrafficLightEstimatorNode::TrafficLightEstimatorNode(const rclcpp::NodeOptions & options)
+: Node("traffic_light_estimator", options)
+{
+  using std::placeholders::_1;
+
+  use_last_detect_color_ = this->declare_parameter("use_last_detect_color", true);
+
+  sub_map_ = create_subscription<HADMapBin>(
+    "~/input/vector_map", rclcpp::QoS{1}.transient_local(),
+    std::bind(&TrafficLightEstimatorNode::onMap, this, _1));
+  sub_route_ = create_subscription<HADMapRoute>(
+    "~/input/route", rclcpp::QoS{1}.transient_local(),
+    std::bind(&TrafficLightEstimatorNode::onRoute, this, _1));
+  sub_traffic_light_array_ = create_subscription<TrafficSignalArray>(
+    "~/input/classified/traffic_signals", rclcpp::QoS{1},
+    std::bind(&TrafficLightEstimatorNode::onTrafficLightArray, this, _1));
+
+  pub_traffic_light_array_ =
+    this->create_publisher<TrafficSignalArray>("~/output/traffic_signals", rclcpp::QoS{1});
+  pub_processing_time_ = std::make_shared<DebugPublisher>(this, "~/debug");
+}
+
+void TrafficLightEstimatorNode::onMap(const HADMapBin::ConstSharedPtr msg)
+{
+  RCLCPP_INFO(get_logger(), "[TrafficLightEstimatorNode]: Start loading lanelet");
+  lanelet_map_ptr_ = std::make_shared<lanelet::LaneletMap>();
+  lanelet::utils::conversion::fromBinMsg(
+    *msg, lanelet_map_ptr_, &traffic_rules_ptr_, &routing_graph_ptr_);
+  const auto traffic_rules = lanelet::traffic_rules::TrafficRulesFactory::create(
+    lanelet::Locations::Germany, lanelet::Participants::Vehicle);
+  const auto pedestrian_rules = lanelet::traffic_rules::TrafficRulesFactory::create(
+    lanelet::Locations::Germany, lanelet::Participants::Pedestrian);
+  lanelet::routing::RoutingGraphConstPtr vehicle_graph =
+    lanelet::routing::RoutingGraph::build(*lanelet_map_ptr_, *traffic_rules);
+  lanelet::routing::RoutingGraphConstPtr pedestrian_graph =
+    lanelet::routing::RoutingGraph::build(*lanelet_map_ptr_, *pedestrian_rules);
+  lanelet::routing::RoutingGraphContainer overall_graphs({vehicle_graph, pedestrian_graph});
+  overall_graphs_ptr_ =
+    std::make_shared<const lanelet::routing::RoutingGraphContainer>(overall_graphs);
+  RCLCPP_INFO(get_logger(), "[TrafficLightEstimatorNode]: Map is loaded");
+}
+
+void TrafficLightEstimatorNode::onRoute(const HADMapRoute::ConstSharedPtr msg)
+{
+  if (lanelet_map_ptr_ == nullptr) {
+    RCLCPP_WARN(get_logger(), "cannot set traffic light in route because don't receive map");
+    return;
+  }
+
+  lanelet::ConstLanelets route_lanelets;
+  for (const auto & segment : msg->segments) {
+    for (const auto & primitive : segment.primitives) {
+      try {
+        route_lanelets.push_back(lanelet_map_ptr_->laneletLayer.get(primitive.id));
+      } catch (const lanelet::NoSuchPrimitiveError & ex) {
+        RCLCPP_ERROR(get_logger(), "%s", ex.what());
+        return;
+      }
+    }
+  }
+
+  conflicting_crosswalks_.clear();
+
+  for (const auto & route_lanelet : route_lanelets) {
+    constexpr int PEDESTRIAN_GRAPH_ID = 1;
```

Comment:
[Q] what is the number? where is defined about the number?

Reply:
I reused the code from following package, but I don't know abot where does the definition exisit :pleading_face: 

https://github.com/autowarefoundation/autoware.universe/blob/ce6ccec0872481c136dc8df8fe4af08992ffee48/planning/behavior_velocity_planner/src/scene_module/crosswalk/manager.cpp#L61-L62

Reply:
I asked the experts of lanelet library about that, the number is index of Participants::Pedestrian in `overall_grapths` and there is no definition in the library.

https://github.com/autowarefoundation/autoware.universe/blob/ce6ccec0872481c136dc8df8fe4af08992ffee48/planning/route_handler/src/route_handler.cpp#L181-L189

Reply:
I see. Thank you